### PR TITLE
ui: Refactor ORT Run creation form

### PR DESCRIPTION
**Pull request by Etsija**
ORT run creation form is rapidly becoming difficult to understand and maintain, so this PR does some crude code refactoring work, mainly to separate job configurations into their own respective components. This will make it easier to add new jobs to the form, for example the Notifier, as requested in #618.

Please see individual commits for details.

### CODE REVIEW

**Comment by mmurto** [comment]
@Etsija if you're writing with the `any`, this example could maybe be helpful: https://codesandbox.io/s/kind-wildflower-1kzege?file=/src/App.tsx.

File: ui/src/components/form/checkbox-field.tsx [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  TName extends FieldPathByValue<TFieldValues, Array<string>>,
+> = {
+  form: UseFormReturn<TFieldValues, TName>;
+  name: TName;
+  label?: string;
+  description?: React.ReactNode;
+  items: readonly { id: string; label: string }[];
+};
+
+export const CheckboxField = <
```
Comment (mmurto) [inline ui/src/components/form/checkbox-field.tsx:51]: This should have a more descriptive name, as `CheckboxField` doesn't convey it being a multiselect.

File: ui/src/components/form/checkbox-field.tsx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                    ? 'indeterminate'
+                    : false
+              }
+              onCheckedChange={(checked) => {
+                const enabledItems = checked
+                  ? items.map((item) => item.id)
+                  : [];
+                form.setValue(
+                  name,
+                  enabledItems as FieldPathValue<TFieldValues, TName>
```
Comment (mmurto) [inline ui/src/components/form/checkbox-field.tsx:85]: This should have a code comment on why `as` can be used here.

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/-components/advisor-fields.tsx [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+              </FormItem>
+            )}
+          />
+          <CheckboxField
+            form={form}
+            name='jobConfigs.advisor.advisors'
+            label='Enabled advisors'
+            description={
+              <>
+                Select the <strong>advisors</strong> enabled for this ORT Run.
```
Comment (mmurto) [inline ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/-components/advisor-fields.tsx:89]: Is this supposed to be `strong`?

File: ui/src/components/form/checkbox-field.tsx [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+type CheckboxFieldProps<
+  TFieldValues extends FieldValues,
+  TName extends FieldPathByValue<TFieldValues, Array<string>>,
+> = {
+  form: UseFormReturn<TFieldValues, TName>;
+  name: TName;
+  label?: string;
+  description?: React.ReactNode;
+  items: readonly { id: string; label: string }[];
```
Comment (mmurto) [inline ui/src/components/form/checkbox-field.tsx:48]: `items` doesn't really tell what it is, maybe `options` would be better?

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/create-run.tsx [hunk 5]
```diff
@@ -99,6 +95,8 @@ const formSchema = z.object({
   jobConfigContext: z.string().optional(),
 });
 
+export type FormValues = ReturnType<typeof formSchema.parse>;
```
Comment (mmurto) [inline ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/create-run.tsx:98]: As this is exported, it should have a descriptive name like `CreateRunFormValues`.

**Review by mmurto — CHANGES_REQUESTED** [review]

File: ui/src/components/form/multi-select-field.tsx [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+                    : false
+              }
+              onCheckedChange={(checked) => {
+                const enabledItems = checked
+                  ? options.map((option) => option.id)
+                  : [];
+                form.setValue(
+                  name,
+                  // Here, we use the `FieldPathValue` type to ensure that the value is an array of strings.
+                  enabledItems as FieldPathValue<TFieldValues, TName>
```
Comment (mmurto) [inline ui/src/components/form/multi-select-field.tsx:90]: Maybe something like:

TypeScript doesn't get this, but `TName extends FieldPathByValue<TFieldValues, Array<string>>`, so the field behind TName is always an `Array<string>` and `options.map((option) => option.id)` is also `Array<string>`, so this type cast is safe.

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/-components/advisor-fields.tsx [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                  />
+                </FormControl>
+              </FormItem>
+            )}
+          />
+          <MultiSelectField
+            form={form}
+            name='jobConfigs.advisor.advisors'
+            label='Enabled advisors'
+            description={<>Select the advisors enabled for this ORT Run.</>}
```
Comment (mmurto) [inline ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/-components/advisor-fields.tsx:87]: Just an info-nit, string can be directly passed as `ReactNode`, so no need to wrap them in Fragment.

**Review by mmurto — CHANGES_REQUESTED** [review]

File: ui/src/components/form/multi-select-field.tsx (hunk 6 above)
Comment (Etsija) [reply ui/src/components/form/multi-select-field.tsx:90]: BTW is there any benefit for using backticks in code comments, or should they just be plain text?

Comment (mmurto) [reply ui/src/components/form/multi-select-field.tsx:90]: To continue, `as` doesn't really ensure anything, but it rather tells TS compiler that whatever it thinks is wrong, here's the correct type. It doesn't do anything to the value, so if you do `const niceNumber = "I'm a number" as number`, you'll get autocompletions and type checking as if it really was a number, but at runtime it's still the same string and can cause bugs.

File: ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/-components/advisor-fields.tsx (hunk 7 above)
Comment (Etsija) [reply ui/src/routes/_layout/organizations/$orgId/products/$productId/repositories/$repoId/-components/advisor-fields.tsx:87]: Yeah I know, but I would like to keep it as it is, to hint the user of the component that this is a ReactNode, so descriptions can include JSX.

File: ui/src/components/form/multi-select-field.tsx (hunk 6 above)
Comment (mmurto) [reply ui/src/components/form/multi-select-field.tsx:90]: > BTW is there any benefit for using backticks in code comments, or should they just be plain text?

Some editors may display them differently, if you put them in docstrings like:

```typescript
/**
 * This is code: `code`
 */
function foo()
```

For regular inline comments probably no difference, just stylistic choice and communicates that it is code.

**Review by mmurto — APPROVED** [review]

**Pull request merged**